resources/assets/js/annotations/ol/PolygonValidator.js

```diff
@@ -0,0 +1,148 @@
+import 'jsts/org/locationtech/jts/monkey'; // This monkey patches jsts prototypes.
+import JstsLinearRing from 'jsts/org/locationtech/jts/geom/LinearRing';
+import JstsPolygon from 'jsts/org/locationtech/jts/geom/Polygon';
+import LinearRing from '@biigle/ol/geom/LinearRing';
+import LineString from '@biigle/ol/geom/LineString';
+import MultiLineString from '@biigle/ol/geom/MultiLineString';
+import MultiPoint from '@biigle/ol/geom/MultiPoint';
+import MultiPolygon from '@biigle/ol/geom/MultiPolygon';
+import OL3Parser from 'jsts/org/locationtech/jts/io/OL3Parser';
+import Point from '@biigle/ol/geom/Point';
+import Polygon from '@biigle/ol/geom/Polygon';
+import Polygonizer from 'jsts/org/locationtech/jts/operation/polygonize/Polygonizer';
+
+/**
+ * Checks if polygon consists of at least 3 unique points
+ *
+ * @param feature containing the polygon
+ * @returns True if coordinates contains at least 3 unique points, otherwise false
+ */
+export function isInvalidPolygon(feature) {
+    let polygon = feature.getGeometry();
+    let points = polygon.getCoordinates()[0];
+
+    return (new Set(points.map(xy => String([xy])))).size < 3;
+}
+
+/**
+ * Makes non-simple polygon simple
+ *
+ * @param feature feature containing the (non-simple) polygon
+ */
+export function makePolygonSimple(feature) {
+    if (feature.getGeometry().getType() !== 'Polygon') {
+        throw new Error("Only polygon geometries are supported.");
+    }
+
+    // Check if polygon is self-intersecting
+    const parser = new OL3Parser();
+    parser.inject(
+        Point,
+        LineString,
+        LinearRing,
+        Polygon,
+        MultiPoint,
+        MultiLineString,
+        MultiPolygon
+    );
+
+    // Translate ol geometry into jsts geometry
+    let jstsPolygon = parser.read(feature.getGeometry());
+
+    if (jstsPolygon.isSimple()) {
+        return feature;
+    }
+
+    // Divide non-simple polygon at cross points into several smaller polygons
+    // and translate back to ol geometry
+    let polygons = jstsValidate(jstsPolygon)['array'].map(p => parser.write(p));
```

Comment:
As a small performance optimization, maybe the greatest polygon can be found *before* converting everything back to OL geometry? Or do you think it's not worth it?

Reply:
Could you explain why this would be an performance optimization?

Reply:
Well, probably it won't have a big effect but my reasoning was this: Here you convert all JSTS Polygons back to OL Polygons (with a "parser" who "writes" which sounds expensive to me :smile:). But right after that you throw away all but one of the converted polygons. So I was thinking that you could throw away the small polygons first before converting them.

Reply:
> maybe the greatest polygon can be found _before_ converting everything back to OL geometry?

Unfortunately this is not possible any more due to the bug mentioned above. There are some cases where polygons were separated and thereby just copied. Some copies still remained in the originals polygons coordinate set which needs to be removed. During the write() process, those coordinates are separated from that original list because they are recognized as different polygons. So I could remove them easily after calling write().
I also tried to catch and handle those case before transforming the polygons but this got really messy. The jsts polygon methods like covers(), contains(), overlaps(), etc... didn't work for that kind of polygons. Additionally I couldn't simply remove these duplicated coordinates because this could still lead to intersecting polygons and cause some other problems.